packages/core/strapi/lib/services/entity-service/types/params/filters/index.ts

```diff
@@ -0,0 +1,75 @@
+import type { Attribute, Common, Utils } from '@strapi/strapi';
+
+import type * as Operator from './operators';
+import type * as AttributeUtils from './attributes';
+
+export { Operator, AttributeUtils as Attribute };
+
+/**
+ * Filter representation for nested attributes
+ */
+type NestedAttributeCondition<
+  TSchemaUID extends Common.UID.Schema,
+  TAttribute extends Attribute.GetKeys<TSchemaUID>
+> = ObjectNotation<
+  Utils.Guard.Never<Attribute.GetTarget<TSchemaUID, TAttribute>, Common.UID.Schema>
+>;
+
+/**
+ * Filter representation for scalar attributes
+ */
+type AttributeCondition<
+  TSchemaUID extends Common.UID.Schema,
+  TAttribute extends Attribute.GetKeys<TSchemaUID>
+> = Utils.Expression.If<
+  Utils.Expression.IsNotNever<TAttribute>,
+  // Get the filter attribute value for the given attribute
+  AttributeUtils.GetValue<Attribute.Get<TSchemaUID, TAttribute>>,
+  // Fallback to the list of all possible scalar attributes' value if the attribute is not valid (never)
+  AttributeUtils.ScalarValues
+> extends infer TAttributeValue
+  ?
+      | TAttributeValue // Implicit $eq operator
+      | ({
+          [TIter in Operator.BooleanValue]?: boolean;
+        } & {
+          [TIter in Operator.DynamicValue]?: TAttributeValue;
+        } & {
+          [TIter in Operator.DynamicArrayValue]?: TAttributeValue[];
+        } & {
+          [TIter in Operator.DynamicBoundValue]?: [TAttributeValue, TAttributeValue];
+        } & {
+          [TIter in Operator.Logical]?: AttributeCondition<TSchemaUID, TAttribute>;
+        } & {
+          [TIter in Operator.Group]?: AttributeCondition<TSchemaUID, TAttribute>[];
+        })
+  : never;
+
+/**
+ * Tree representation of a Strapi filter for a given schema UID
+ */
+export type ObjectNotation<TSchemaUID extends Common.UID.Schema> = {
+  [TIter in Operator.Group]?: ObjectNotation<TSchemaUID>[];
+} & {
+  [TIter in Operator.Logical]?: ObjectNotation<TSchemaUID>;
+} & ([AttributeUtils.GetScalarKeys<TSchemaUID>, AttributeUtils.GetNestedKeys<TSchemaUID>] extends [
+    infer TScalarKeys extends AttributeUtils.GetScalarKeys<TSchemaUID>,
+    infer TNestedKeys extends AttributeUtils.GetNestedKeys<TSchemaUID>
+  ]
+    ? // If both the scalar and nested keys are resolved, then create a strongly
+      // typed filter object, else create a loose generic abstraction.
+      Utils.Expression.If<
+        Utils.Expression.And<
+          Utils.Expression.IsNotNever<TScalarKeys>,
+          Utils.Expression.IsNotNever<TNestedKeys>
```

Comment:
Do you think it could be more readable if we imported the expression directly?
```suggestion
      If<
        And<
          IsNotNever<TScalarKeys>,
          IsNotNever<TNestedKeys>
```
( I don't see this mentioned in the Draft TS guidelines )

Reply:
We were talking about that with Alex earlier today.

I would personally advocate for keeping the namespaces even if it adds some verbosity.
However, we thought about adding simpler aliases for every namespaces so that we can have shorter lines (maybe smth like U.E.If or smth else)

What do you think?

Reply:
Also, it does indeed makes sense to add smth in the guidelines about that

Reply:
Aliases would work for me :) sounds like a good compromise

Reply:
I think there was a debate on whether we should or shouldn't have a namespace for the utils